### ISSUE TITLE
use structs to mimic default arguments

### ISSUE THREAD

**Issue by Kiyoshika** [Kiyoshika/Generalized-Model-Framework#10]
due to lack of default arguments in C, the parameters for the optimizers are getting very large.
Would like to mimic default parameters using structs.

E.g. there could be a `gradient_descent_params` struct which establishes default values for function pointers (NULL) and other values (iterations, learning rate, etc.)

The user can create their own custom parameters using initializers and wouldn't have to specify every parameter manually. Something such as

```c
// here the user can specify only the parameters they care about changing.
// every other parameter would be set to some default value 
// even the weights parameter can be stored internally rather than being initialized by user. However,
// this would require to implement a separate free() function to free the memory allocated to the weights.
gradient_descent_params my_model = 
{
    .data = data,
    .data_dims = data_dims,
    .target = target,
    .activation = &gmf_activation_linear,
    .loss = &gmf_loss_squared
};
```

**Comment by Kiyoshika** [comment]
implemented in 7045cc01f49e170ca808246277409dc1a4c8439e
